### ISSUE TITLE
In QgsMapLayerComboBox the order of the layers is reverse (or even mixed?) Bug or Feature?

### ISSUE THREAD

**Issue by rduivenvoorde** [qgis/QGIS#47047]
### What is the bug or the crash?

When creating a plugin and using QgsMapLayerComboBox, I see that the order in the combobox is reversed compared to the layer list:

![Screenshot-20220128124446-1022x232](https://user-images.githubusercontent.com/731673/151541970-6842a3c3-baa4-4156-bc5c-f355a082681c.png)

Is there a specific reason for it? 

In my case the user has to select a layer, and it is counterintuitive if you have to select your top layer from the bottom?

It looks like a trival fix (but... you never know...) but I want to be sure before I would do/try a PR that I do not break others use of the QgsMapLayerComboBox

### Steps to reproduce the issue

Use the following snippet to have a dialog showing a QgsMapLayerComboBox of your current project

```
d = QDialog(iface.mainWindow())
# mmm, tried a QgsDialog here but then the layout does not show up???
#d = QgsDialog(iface.mainWindow())

d.resize(300,300)

layout = QVBoxLayout()

layer_combo = QgsMapLayerComboBox()
layout.addWidget(layer_combo)

layout.addWidget(QPushButton("Test"))
d.setLayout(layout)

d.show()
```

Mmm, maybe it is even more complex??? Apparently it is not only revers, but actually shuffled too? My bottom opentopo reference layer is in the middle?

![image](https://user-images.githubusercontent.com/731673/151544952-438d7863-c4fb-4e04-9162-e94513319665.png)

### Versions

QGIS version | 3.23.0-Master | QGIS code revision | 31ac4082f84
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.10
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Debian GNU/Linux bookworm/sis

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by nicogodet** [comment]
It looks ordered in alphabetical order to me.

**Comment by rduivenvoorde** [comment]
@nicogodet mmm, you are probably right :-)

**Comment by rduivenvoorde** [comment]
https://github.com/qgis/QGIS/blob/master/src/core/qgsmaplayerproxymodel.cpp#L36

That is the model that is behind de ComboBox... so indeed it IS sorted...

So my main question then: is this preferable? 

Or should I try to add an option to NOT sort it? 

Or is the order in such models undefined? (don't think so, because normal comboboxes you can fill with an array of data, and that order is preserved)?

**Comment by roya0045** [comment]
Some UI elements partly sort by alphabetical order and then just add the remaining item (unless the layer list has recently changed). 

I think having al alphabetical sort helps to find where elements are going to be in the list as it is reliable. Using the layertree order might be ok but will possibly require more time parsing the list as opposed to the current sorting.

**Comment by 3nids** [comment]
@rduivenvoorde I think the commit you point was fixing an issue with accent or something similar, not sure.

It's true that if you have a flat layer tree with a couple of layers, it would make more sense to have them in the same order.
But as soon as the list grows and that it gets grouped, the alphabetical way is the only way to go I think (except if you have several layers with same name).

I think there would be two possible improvements to this widget:
1. make it searchable by typing text
2. have a button to show a layer tree to choose the layer (I think the model used for snapping or DXF export could be reused for this)

**Comment by rduivenvoorde** [comment]
@3nids yeah, you are probably right, I did not even think about grouped layers etc etc...

**Comment by alexbruy** [comment]
@rduivenvoorde should turn it into a feature request and change title?

**Comment by rduivenvoorde** [comment]
@alexbruy will do. I think the layer tree would be the best option? So it fit's the same visual looks as the layer tree itself?
(so in case you have several layers with the same name, you can still choose from the context)?

A search box on top (to search in a treeview then) would be an extra feature
